workspace/__ardulib__/WS2812_NoBufffer/WS2812.cpp

```diff
@@ -0,0 +1,152 @@
+#include "WS2812.h"
+
+WS2812::WS2812(uint16_t pin, uint32_t len) {
+    _length = len;
+    setup(pin);
+}
+
+int8_t getBitPosition(uint8_t b) {
+    int8_t pos = -1;
+    uint8_t t = b;
+    while (t > 0) {
+        t >>= 1;
+        pos++;
+    }
+    return pos;
+}
+
+// Set the specified pin up as digital out
+void WS2812::setup(uint16_t pin) {
+    pinMode(pin, OUTPUT);
+    digitalWrite(pin, LOW);
+    _port = portOutputRegister(digitalPinToPort(pin));
+#if defined(ESP8266) || defined(ESP32)
+    _bit = pin;
+#else
+    _bit = getBitPosition(digitalPinToBitMask(pin));
+#endif
```

Comment:
The magic should be annotated.

I suspect calling it `_bit` in the ESP case is plainly wrong and misguiding.